The problem seems to be that `buildbot` runs tests that require `buildbot-pkg` to be available. `buildbot-pkg` is currently only used by the plugins, and is not a real package (the derivation is defined in a `let` block in `plugins.nix`).

I moved `buildbot-pkg` to its own file and added it as a `buildInput` of buildbot. This fixed the tests, but introduced a new problem, seemingly related to how buildbot_pkg monkey patches `os.listdir`

This is the current error, which I can't find a way to fix:
```
Traceback (most recent call last):
  File "nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 556, in <module>
    setup(**setup_args)
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/nix/store/87l8l9w3x8saxqz9sfcfw2i3jiilmxfi-python-2.7.14/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/nix/store/87l8l9w3x8saxqz9sfcfw2i3jiilmxfi-python-2.7.14/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/nix/store/87l8l9w3x8saxqz9sfcfw2i3jiilmxfi-python-2.7.14/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/setuptools/command/test.py", line 226, in run
    self.run_tests()
  File "/nix/store/87l8l9w3x8saxqz9sfcfw2i3jiilmxfi-python-2.7.14/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/setuptools/command/test.py", line 171, in project_on_sys_path
    working_set.__init__()
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/pkg_resources/__init__.py", line 650, in __init__
    self.add_entry(entry)
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/pkg_resources/__init__.py", line 706, in add_entry
    for dist in find_distributions(entry, True):
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2032, in find_on_path
    entries = safe_listdir(path_item)
  File "/nix/store/a4zb290zkx8avpand4a58ndci8ppczik-python2.7-bootstrapped-pip-9.0.1/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2091, in safe_listdir
    return os.listdir(path)
  File "/nix/store/8gkpajpp6fl78ck3jjdy3znnpswqakm4-python2.7-buildbot-pkg-0.9.15.post1/lib/python2.7/site-packages/buildbot_pkg.py", line 41, in listdir
    l = old_listdir(path)
TypeError: 'NoneType' object is not callable
```

--- a/pkgs/development/tools/build-managers/buildbot/default.nix
+++ b/pkgs/development/tools/build-managers/buildbot/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, openssh, buildbot-worker, pythonPackages, runCommand, makeWrapper }:
+{ stdenv, lib, openssh, buildbot-worker, buildbot-pkg, pythonPackages, runCommand, makeWrapper }:
 
 let
   withPlugins = plugins: runCommand "wrapped-${package.name}" {
@@ -36,6 +36,7 @@ let
       pyflakes
       openssh
       buildbot-worker
+      buildbot-pkg
       treq
     ];
 

--- a/pkgs/development/tools/build-managers/buildbot/pkg.nix
+++ b/pkgs/development/tools/build-managers/buildbot/pkg.nix
@@ -1,0 +1,21 @@
+{ stdenv, buildPythonPackage, fetchPypi, setuptools }:
+
+buildPythonPackage rec {
+  name = "${pname}-${version}";
+  pname = "buildbot-pkg";
+  version = "0.9.15.post1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0gsa5fi1gkwnz8dsrl2s5kzcfawnj3nl8g8h6z1winz627l9n8sh";
+  };
+
+  propagatedBuildInputs = [ setuptools ];
+
+  meta = with stdenv.lib; {
+    homepage = http://buildbot.net/;
+    description = "Buildbot Packaging Helper";
+    maintainers = with maintainers; [ nand0p ryansydnor ];
+    license = licenses.gpl2;
+  };
+}

--- a/pkgs/development/tools/build-managers/buildbot/plugins.nix
+++ b/pkgs/development/tools/build-managers/buildbot/plugins.nix
@@ -1,27 +1,6 @@
-{ stdenv, pythonPackages }:
+{ stdenv, pythonPackages, buildbot-pkg }:
 
-let
-  buildbot-pkg = pythonPackages.buildPythonPackage rec {
-    name = "${pname}-${version}";
-    pname = "buildbot-pkg";
-    version = "0.9.15.post1";
-
-    src = pythonPackages.fetchPypi {
-      inherit pname version;
-      sha256 = "0gsa5fi1gkwnz8dsrl2s5kzcfawnj3nl8g8h6z1winz627l9n8sh";
-    };
-
-    propagatedBuildInputs = with pythonPackages; [ setuptools ];
-
-    meta = with stdenv.lib; {
-      homepage = http://buildbot.net/;
-      description = "Buildbot Packaging Helper";
-      maintainers = with maintainers; [ nand0p ryansydnor ];
-      license = licenses.gpl2;
-    };
-  };
-
-in {
+{
   www = pythonPackages.buildPythonPackage rec {
     name = "${pname}-${version}";
     pname = "buildbot_www";

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -7267,6 +7267,9 @@ with pkgs;
   buildbot-worker = callPackage ../development/tools/build-managers/buildbot/worker.nix {
     pythonPackages = python2Packages;
   };
+  buildbot-pkg = callPackage ../development/tools/build-managers/buildbot/pkg.nix {
+    inherit (python2Packages) buildPythonPackage fetchPypi setuptools;
+  };
   buildbot-plugins = callPackages ../development/tools/build-managers/buildbot/plugins.nix {
     pythonPackages = python2Packages;
   };

